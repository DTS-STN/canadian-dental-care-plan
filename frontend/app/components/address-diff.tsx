import type { ComponentProps } from 'react';

import { DiffViewer } from './diff-viewer';
import { formatAddress } from '~/utils/string-utils';
import { cn } from '~/utils/tw-utils';

interface Address {
  address: string;
  apartment?: string;
  city: string;
  country: string;
  postalZipCode?: string;
  provinceState?: string;
}

export interface AddressDiffProps extends ComponentProps<'address'> {
  altFormat?: boolean;
  newAddress: Address;
  oldAddress: Address;
}

export function AddressDiff({ altFormat, oldAddress, newAddress, className, ...restProps }: AddressDiffProps) {
  const formattedOldAddress = formatAddress({ ...oldAddress, altFormat });
  const formattedNewAddress = formatAddress({ ...newAddress, altFormat });
  return (
    <address className={cn('whitespace-pre-wrap not-italic', className)} data-testid="address-diff" {...restProps}>
      <DiffViewer oldStr={formattedOldAddress} newStr={formattedNewAddress} />
    </address>
  );
}
