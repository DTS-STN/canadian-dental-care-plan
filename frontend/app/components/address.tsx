import type { ComponentProps } from 'react';

import clsx from 'clsx';

export interface AddressProps extends ComponentProps<'address'> {
  address: string;
  city: string;
  provinceState?: string;
  postalZipCode?: string;
  country: string;
}

function formatAddress(address: string, city: string, country: string, provinceState?: string, postalZipCode?: string) {
  // TODO 'canada' shouldn't be hardcoded as we may deal with different values of the country field such as abbreviations
  const isNotCanadianAddress = 'canada' !== country.toLowerCase();

  // prettier-ignore
  const lines = [`${address}`, 
    `${city}${provinceState ? ` ${provinceState}` : ''}${postalZipCode ? `  ${postalZipCode}` : ''}`, 
    `${isNotCanadianAddress ? country : ''}`];

  return lines
    .map((line) => line.trim())
    .filter(Boolean)
    .join('\n');
}

export function Address(props: AddressProps) {
  const { address, city, provinceState, postalZipCode, country, className, ...restProps } = props;
  const formattedAddress = formatAddress(address, city, country, provinceState, postalZipCode);

  return (
    <address className={clsx('mb-0 whitespace-pre-wrap', className)} data-testid="address-id" {...restProps}>
      {formattedAddress}
    </address>
  );
}
