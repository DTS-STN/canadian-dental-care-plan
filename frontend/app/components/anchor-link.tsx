import type { ComponentProps, MouseEvent } from 'react';

import { scrollAndFocusFromAnchorLink } from '~/utils/link-utils';

/**
 * AnchorLinkProps represents the properties for the AnchorLink component.
 * It extends the ComponentProps<'a'> type, omitting the 'href' property,
 * and adds the required 'anchorElementId' property.
 */
export interface AnchorLinkProps extends OmitStrict<ComponentProps<'a'>, 'href'> {
  anchorElementId: string;
}

/**
 * AnchorLink is a React component used to create anchor links that scroll
 * and focus on the specified target element when clicked.
 *
 * @param {AnchorLinkProps} props - The properties for the AnchorLink component.
 * @returns {JSX.Element} JSX element representing the anchor link.
 */
export function AnchorLink(props: AnchorLinkProps) {
  const { anchorElementId, children, onClick, ...restProps } = props;

  /**
   * handleOnSkipLinkClick is the click event handler for the anchor link.
   * It prevents the default anchor link behavior, scrolls to and focuses
   * on the target element specified by 'anchorElementId', and invokes
   * the optional 'onClick' callback.
   */
  function handleOnSkipLinkClick(e: MouseEvent<HTMLAnchorElement>) {
    e.preventDefault();
    scrollAndFocusFromAnchorLink(e.currentTarget.href);
    onClick?.(e);
  }

  return (
    <a href={`#${anchorElementId}`} onClick={handleOnSkipLinkClick} data-testid="anchor-link" {...restProps}>
      {children}
    </a>
  );
}
