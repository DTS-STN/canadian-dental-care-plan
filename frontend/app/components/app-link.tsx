import { Link, type LinkProps } from '@remix-run/react';
import { useTranslation } from 'react-i18next';

/**
 * Props for the AppLink component.
 */
export type AppLinkProps = { locale?: 'en' | 'fr' } & LinkProps;

/**
 * A component that renders a localized link.
 */
export function AppLink({ locale, ...props }: AppLinkProps) {
  const { i18n } = useTranslation();

  const targetLocale = locale ?? i18n.language;
  const path = `/${targetLocale}${props.to}`;

  return (
    <Link {...props} to={path}>
      {props.children}
    </Link>
  );
}
