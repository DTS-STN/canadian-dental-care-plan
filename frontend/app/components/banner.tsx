import type { ReactNode } from 'react';

interface BannerProps {
  alert?: ReactNode;
  description?: ReactNode;
}

export function Banner({ alert, description }: BannerProps) {
  return (
    <div className="font-body bg-slate-700 text-white" data-testid="banner-id">
      <div className="container mx-auto flex flex-col space-y-2 p-4 lg:flex-row lg:items-center lg:space-y-0 lg:space-x-4">
        <div className="w-max border-2 border-current px-4 py-1 whitespace-nowrap">{alert}</div>
        <div>{description}</div>
      </div>
    </div>
  );
}
