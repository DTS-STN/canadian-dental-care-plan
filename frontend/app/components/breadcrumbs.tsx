import { ReactNode } from 'react';

import { faChevronRight } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useTranslation } from 'react-i18next';
import type { To } from 'react-router';

import { InlineLink } from './inline-link';

export interface BreadcrumbsProps {
  items: Array<{ content: string; to?: To }>;
}

export function Breadcrumbs({ items }: BreadcrumbsProps) {
  const { t } = useTranslation(['gcweb']);
  return (
    <nav id="wb-bc" property="breadcrumb" aria-labelledby="breadcrumbs">
      <h2 id="breadcrumbs" className="sr-only">
        {t('gcweb:breadcrumbs.you-are-here')}
      </h2>
      <div className="container mt-4">
        <ol className="flex flex-wrap items-center gap-x-3 gap-y-1" typeof="BreadcrumbList">
          {items.map(({ content, to }, idx) => {
            return (
              <li key={content} property="itemListElement" typeof="ListItem" className="flex items-center">
                {idx !== 0 && <FontAwesomeIcon icon={faChevronRight} className="mr-2 size-3 text-slate-700" />}
                <Breadcrumb to={to}>{content}</Breadcrumb>
              </li>
            );
          })}
        </ol>
      </div>
    </nav>
  );
}

function Breadcrumb({ children, to }: { children: ReactNode; to?: To }) {
  // prettier-ignore
  return to === undefined
    ? <span property="name">{children}</span>
    : <InlineLink to={to} property="item" typeof="WebPage"><span property="name">{children}</span></InlineLink>;
}
