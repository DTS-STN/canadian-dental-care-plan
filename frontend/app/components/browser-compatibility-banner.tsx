import type { JSX } from 'react';

import { faCircleExclamation } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useTranslation } from 'react-i18next';

import { Button } from '~/components/buttons';

interface BrowserCompatibilityBannerProps {
  onDismiss?: () => void;
}

export function BrowserCompatibilityBanner({ onDismiss }: BrowserCompatibilityBannerProps): JSX.Element | undefined {
  const { t } = useTranslation(['gcweb']);

  return (
    <div id="browser-compatibility-banner" className="border-b border-red-700 bg-red-50" role="alert">
      <div className="container">
        <div className="flex items-stretch">
          <div className="hidden items-center bg-red-200 p-4 sm:flex">
            <FontAwesomeIcon icon={faCircleExclamation} className="size-6 shrink-0 text-red-700" />
          </div>
          <div className="p-4">
            <h2 className="font-lato mb-1 font-semibold">{t('gcweb:browser-compatibility-banner.title')}</h2>
            <p>{t('gcweb:browser-compatibility-banner.content')}</p>
            <div className="text-right">
              <Button variant="link" type="button" onClick={onDismiss}>
                {t('gcweb:browser-compatibility-banner.dismiss')}
              </Button>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}
