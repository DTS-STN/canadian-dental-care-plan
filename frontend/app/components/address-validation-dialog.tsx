import { useState } from 'react';
import type { SyntheticEvent } from 'react';

import { faChevronLeft, faChevronRight } from '@fortawesome/free-solid-svg-icons';
import { useTranslation } from 'react-i18next';
import invariant from 'tiny-invariant';

import { Address } from '~/components/address';
import { Button } from '~/components/buttons';
import { DialogClose, DialogContent, DialogDescription, DialogFooter, DialogHeader, DialogTitle } from '~/components/dialog';
import { InputRadios } from '~/components/input-radios';
import { LoadingButton } from '~/components/loading-button';
import { useEnhancedFetcher } from '~/hooks';

export interface CanadianAddress {
  address: string;
  city: string;
  country: string;
  countryId: string;
  postalZipCode: string;
  provinceState: string;
  provinceStateId: string;
}

export interface AddressSuggestionResponse {
  enteredAddress: CanadianAddress;
  status: 'address-suggestion';
  suggestedAddress: CanadianAddress;
}

export interface AddressInvalidResponse {
  invalidAddress: CanadianAddress;
  status: 'address-invalid';
}

export type AddressResponse = AddressSuggestionResponse | AddressInvalidResponse;

interface AddressSuggestionDialogContentProps {
  enteredAddress: CanadianAddress;
  suggestedAddress: CanadianAddress;
  formAction: string;
  syncAddresses?: boolean;
}

export function AddressSuggestionDialogContent({ enteredAddress, suggestedAddress, formAction, syncAddresses = false }: AddressSuggestionDialogContentProps) {
  const { t } = useTranslation(['common']);
  const fetcher = useEnhancedFetcher();
  const enteredAddressOptionValue = 'entered-address';
  const suggestedAddressOptionValue = 'suggested-address';
  type AddressSelectionOption = typeof enteredAddressOptionValue | typeof suggestedAddressOptionValue;
  const [selectedAddressSuggestionOption, setSelectedAddressSuggestionOption] = useState<AddressSelectionOption>(suggestedAddressOptionValue);

  async function onSubmitHandler(event: SyntheticEvent<HTMLFormElement, SubmitEvent>) {
    event.preventDefault();
    const formData = new FormData(event.currentTarget);

    // Get the clicked button's value and append it to the FormData object
    const submitter = event.nativeEvent.submitter as HTMLButtonElement | null;
    invariant(submitter, 'Expected submitter to be defined');
    formData.set(submitter.name, submitter.value);

    // Append selected address suggestion to form data
    const selectedAddressSuggestion = selectedAddressSuggestionOption === enteredAddressOptionValue ? enteredAddress : suggestedAddress;
    formData.set('address', selectedAddressSuggestion.address);
    formData.set('city', selectedAddressSuggestion.city);
    formData.set('countryId', selectedAddressSuggestion.countryId);
    formData.set('postalZipCode', selectedAddressSuggestion.postalZipCode);
    formData.set('provinceStateId', selectedAddressSuggestion.provinceStateId);
    if (syncAddresses) {
      formData.set('syncAddresses', 'true');
    }

    await fetcher.submit(formData, { method: 'POST' });
  }

  return (
    <DialogContent aria-describedby={undefined} className="sm:max-w-md">
      <DialogHeader>
        <DialogTitle>{t('common:dialog.address-suggestion.header')}</DialogTitle>
        <DialogDescription id="verify-description">{t('common:dialog.address-suggestion.description')}</DialogDescription>
      </DialogHeader>
      <InputRadios
        id="addressSelection"
        name="addressSelection"
        legend={t('common:dialog.address-suggestion.address-selection-legend')}
        outerAriaDescribedById="verify-description"
        options={[
          {
            value: enteredAddressOptionValue,
            children: (
              <>
                <p className="mb-2">
                  <strong>{t('common:dialog.address-suggestion.entered-address-option')}</strong>
                </p>
                <Address address={enteredAddress} />
              </>
            ),
          },
          {
            value: suggestedAddressOptionValue,
            children: (
              <>
                <p className="mb-2">
                  <strong>{t('common:dialog.address-suggestion.suggested-address-option')}</strong>
                </p>
                <Address address={suggestedAddress} />
              </>
            ),
          },
        ].map((option) => ({
          ...option,
          onChange: (e) => {
            setSelectedAddressSuggestionOption(e.target.value as AddressSelectionOption);
          },
          checked: option.value === selectedAddressSuggestionOption,
        }))}
      />
      <DialogFooter>
        <DialogClose asChild>
          <Button
            id="dialog.corrected-address-close-button"
            disabled={fetcher.isSubmitting}
            variant="default"
            size="sm"
            startIcon={faChevronLeft}
            data-gc-analytics-customclick="ESDC-EDSC:CDCP Online Application Form:Dialog Back - Address Suggestion click"
          >
            {t('common:dialog.address-suggestion.cancel-button')}
          </Button>
        </DialogClose>
        <fetcher.Form method="post" noValidate onSubmit={onSubmitHandler}>
          <LoadingButton
            name="_action"
            value={formAction}
            type="submit"
            id="dialog.corrected-address-use-selected-address-button"
            loading={fetcher.isSubmitting}
            endIcon={faChevronRight}
            variant="primary"
            size="sm"
            data-gc-analytics-customclick="ESDC-EDSC:CDCP Online Application Form:Dialog Use Selected Address - Address Suggestion click"
          >
            {t('common:dialog.address-suggestion.use-selected-address-button')}
          </LoadingButton>
        </fetcher.Form>
      </DialogFooter>
    </DialogContent>
  );
}

interface AddressInvalidDialogContentProps {
  invalidAddress: CanadianAddress;
  formAction: string;
  syncAddresses?: boolean;
}

export function AddressInvalidDialogContent({ formAction, invalidAddress, syncAddresses = false }: AddressInvalidDialogContentProps) {
  const { t } = useTranslation(['common']);
  const fetcher = useEnhancedFetcher();

  async function onSubmitHandler(event: SyntheticEvent<HTMLFormElement, SubmitEvent>) {
    event.preventDefault();
    const formData = new FormData(event.currentTarget);

    // Get the clicked button's value and append it to the FormData object
    const submitter = event.nativeEvent.submitter as HTMLButtonElement | null;
    invariant(submitter, 'Expected submitter to be defined');
    formData.set(submitter.name, submitter.value);

    // Append selected address suggestion to form data
    formData.set('address', invalidAddress.address);
    formData.set('city', invalidAddress.city);
    formData.set('countryId', invalidAddress.countryId);
    formData.set('postalZipCode', invalidAddress.postalZipCode);
    formData.set('provinceStateId', invalidAddress.provinceStateId);
    if (syncAddresses) {
      formData.set('syncAddresses', 'true');
    }

    await fetcher.submit(formData, { method: 'POST' });
  }

  return (
    <DialogContent aria-describedby={undefined} className="sm:max-w-md">
      <DialogHeader>
        <DialogTitle>{t('common:dialog.address-invalid.header')}</DialogTitle>
        <DialogDescription>{t('common:dialog.address-invalid.description')}</DialogDescription>
      </DialogHeader>
      <div className="space-y-2">
        <p>
          <strong>{t('common:dialog.address-invalid.entered-address')}</strong>
        </p>
        <Address address={invalidAddress} />
      </div>
      <DialogFooter>
        <DialogClose asChild>
          <Button id="dialog.address-invalid-close-button" variant="default" size="sm" startIcon={faChevronLeft} data-gc-analytics-customclick="ESDC-EDSC:CDCP Online Application Form:Dialog Back - Address Invalid click">
            {t('common:dialog.address-invalid.close-button')}
          </Button>
        </DialogClose>
        <fetcher.Form method="post" noValidate onSubmit={onSubmitHandler}>
          <LoadingButton
            name="_action"
            value={formAction}
            type="submit"
            id="dialog.address-invalid-use-entered-address-button"
            loading={fetcher.isSubmitting}
            endIcon={faChevronRight}
            variant="primary"
            size="sm"
            data-gc-analytics-customclick="ESDC-EDSC:CDCP Online Application Form:Dialog Use entered address - Address Invalid click"
          >
            {t('common:dialog.address-invalid.use-entered-address-button')}
          </LoadingButton>
        </fetcher.Form>
      </DialogFooter>
    </DialogContent>
  );
}
