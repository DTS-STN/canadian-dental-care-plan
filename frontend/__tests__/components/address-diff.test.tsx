import { render } from '@testing-library/react';

import { describe, expect, it } from 'vitest';

import { AddressDiff } from '~/components/address-diff';

describe('AddressDiff', () => {
  it('should render the component', () => {
    const oldAddress = {
      address: '123 Main St',
      city: 'Anytown',
      provinceState: 'ON',
      postalZipCode: 'K1A 0B1',
      country: 'Canada',
      apartment: 'Apt 4',
    };

    const newAddress = {
      address: '456 Oak Ave',
      city: 'Otherville',
      provinceState: 'QC',
      postalZipCode: 'H2B 2C3',
      country: 'Canada',
      apartment: 'Unit 5',
    };

    const { container } = render(<AddressDiff oldAddress={oldAddress} newAddress={newAddress} />);
    expect(container).toMatchSnapshot();
  });

  it('should render the component with an alternate format', () => {
    const oldAddress = {
      address: '123 Main St',
      city: 'Anytown',
      provinceState: 'ON',
      postalZipCode: 'K1A 0B1',
      country: 'Canada',
      apartment: 'Apt 4',
    };

    const newAddress = {
      address: '456 Oak Ave',
      city: 'Otherville',
      provinceState: 'QC',
      postalZipCode: 'H2B 2C3',
      country: 'Canada',
      apartment: 'Unit 5',
    };

    const { container } = render(<AddressDiff oldAddress={oldAddress} newAddress={newAddress} altFormat />);
    expect(container).toMatchSnapshot();
  });
});
