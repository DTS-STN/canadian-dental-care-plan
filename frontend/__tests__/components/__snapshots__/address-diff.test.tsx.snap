// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddressDiff > should render the component 1`] = `
<div>
  <address
    class="whitespace-pre-wrap not-italic"
    data-testid="address-diff"
  >
    <div
      class="diff-viewer-content"
    >
      <del
        class="bg-red-100"
      >
        123
      </del>
      <ins
        class="bg-green-100"
      >
        456
      </ins>
       
      <del
        class="bg-red-100"
      >
        M
      </del>
      <ins
        class="bg-green-100"
      >
        O
      </ins>
      a
      <del
        class="bg-red-100"
      >
        in
      </del>
      <ins
        class="bg-green-100"
      >
        k
      </ins>
       
      <del
        class="bg-red-100"
      >
        St 
      </del>
      A
      <del
        class="bg-red-100"
      >
        pt
      </del>
      <ins
        class="bg-green-100"
      >
        ve
      </ins>
       
      <del
        class="bg-red-100"
      >
        4
A
      </del>
      <ins
        class="bg-green-100"
      >
        U
      </ins>
      n
      <del
        class="bg-red-100"
      >
        y
      </del>
      <ins
        class="bg-green-100"
      >
        i
      </ins>
      t
      <del
        class="bg-red-100"
      >
        own
      </del>
       
      <ins
        class="bg-green-100"
      >
        5

      </ins>
      O
      <del
        class="bg-red-100"
      >
        N
      </del>
      <ins
        class="bg-green-100"
      >
        therville
      </ins>
       
      <ins
        class="bg-green-100"
      >
        QC
      </ins>
       
      <del
        class="bg-red-100"
      >
        K1A
      </del>
       
      <del
        class="bg-red-100"
      >
        0
      </del>
      <ins
        class="bg-green-100"
      >
        H2
      </ins>
      B
      <del
        class="bg-red-100"
      >
        1
      </del>
      <ins
        class="bg-green-100"
      >
         2C3
      </ins>
      
Canada
    </div>
  </address>
</div>
`;

exports[`AddressDiff > should render the component with an alternate format 1`] = `
<div>
  <address
    class="whitespace-pre-wrap not-italic"
    data-testid="address-diff"
  >
    <div
      class="diff-viewer-content"
    >
      <del
        class="bg-red-100"
      >
        123
      </del>
      <ins
        class="bg-green-100"
      >
        456
      </ins>
       
      <del
        class="bg-red-100"
      >
        M
      </del>
      <ins
        class="bg-green-100"
      >
        O
      </ins>
      a
      <del
        class="bg-red-100"
      >
        in
      </del>
      <ins
        class="bg-green-100"
      >
        k
      </ins>
       
      <del
        class="bg-red-100"
      >
        St 
      </del>
      A
      <del
        class="bg-red-100"
      >
        pt
      </del>
      <ins
        class="bg-green-100"
      >
        ve
      </ins>
       
      <del
        class="bg-red-100"
      >
        4
A
      </del>
      <ins
        class="bg-green-100"
      >
        U
      </ins>
      n
      <del
        class="bg-red-100"
      >
        y
      </del>
      <ins
        class="bg-green-100"
      >
        i
      </ins>
      t
      <ins
        class="bg-green-100"
      >
         5

      </ins>
      O
      <del
        class="bg-red-100"
      >
        wn
      </del>
      <ins
        class="bg-green-100"
      >
        therville
      </ins>
      , 
      <del
        class="bg-red-100"
      >
        ON
      </del>
      <ins
        class="bg-green-100"
      >
        QC
      </ins>
      

      <del
        class="bg-red-100"
      >
        K1A 0
      </del>
      <ins
        class="bg-green-100"
      >
        H2
      </ins>
      B
      <del
        class="bg-red-100"
      >
        1
      </del>
      <ins
        class="bg-green-100"
      >
         2C3
      </ins>
      
Canada
    </div>
  </address>
</div>
`;
